ddtrace/tracer/sampler.go

```diff
@@ -147,3 +154,249 @@ func (ps *prioritySampler) apply(spn *span) {
+	rate := sampleRate()
+	return &rulesSampler{
+		rules:   appliedSamplingRules(rules),
+		rate:    rate,
+		limiter: newRateLimiter(rate),
+		ts:      time.Now().Truncate(time.Second),
+	}
+}
+
+// appliedSamplingRules validates the user-provided rules and returns an internal representation.
+// If the DD_TRACE_SAMPLING_RULES environment variable is set, it will replace the given rules.
+func appliedSamplingRules(rules []SamplingRule) []SamplingRule {
+	rulesFromEnv := os.Getenv("DD_TRACE_SAMPLING_RULES")
+	if rulesFromEnv != "" {
+		rules = rules[:0]
+		jsonRules := []struct {
+			Service string      `json:"service"`
+			Name    string      `json:"name"`
+			Rate    json.Number `json:"rate"`
+		}{}
+		err := json.Unmarshal([]byte(rulesFromEnv), &jsonRules)
+		if err != nil {
+			log.Warn("error parsing DD_TRACE_SAMPLING_RULES: %v", err)
+			return nil
+		}
+		for _, v := range jsonRules {
+			if v.Rate == "" {
+				log.Warn("error parsing rule: rate not provided")
+				continue
+			}
+			rate, err := v.Rate.Float64()
+			if err != nil {
+				log.Warn("error parsing rule: invalid rate: %v", err)
+				continue
+			}
+			switch {
+			case v.Service != "" && v.Name != "":
+				rules = append(rules, NameServiceRule(v.Name, v.Service, rate))
+			case v.Service != "":
+				rules = append(rules, ServiceRule(v.Service, rate))
+			case v.Name != "":
+				rules = append(rules, NameRule(v.Name, rate))
+			}
+		}
+	}
+	validRules := make([]SamplingRule, 0, len(rules))
+	for _, v := range rules {
+		if !(v.Rate >= 0.0 && v.Rate <= 1.0) {
+			log.Warn("ignoring rule %+v: rate is out of range", v)
+			continue
+		}
+		validRules = append(validRules, v)
+	}
+	return validRules
+}
+
+// sampleRate returns the sampling rate found in the DD_TRACE_SAMPLE_RATE environment variable. If it is invalid and
+// not within the 0-1 range, the default value of 0 is returned.
+func sampleRate() float64 {
+	defaultRate := math.NaN()
+	v := os.Getenv("DD_TRACE_SAMPLE_RATE")
+	if v == "" {
+		return defaultRate
+	}
+	r, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("ignoring DD_TRACE_SAMPLE_RATE: error: %v", err)
+		return defaultRate
+	}
+	if r >= 0.0 && r <= 1.0 {
+		return r
+	}
+	log.Warn("ignoring DD_TRACE_SAMPLE_RATE: out of range %f", r)
+	return defaultRate
+}
+
+// newRateLimiter returns a new rate limiter which allows r events per second.
+// The DD_TRACE_LIMIT environment variable may override r.
+func newRateLimiter(r float64) *rate.Limiter {
+	defaultLimiter := rate.NewLimiter(rate.Inf, 0)
+	v := os.Getenv("DD_TRACE_RATE_LIMIT")
+	if v == "" {
+		return defaultLimiter
+	}
+	l, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("using default rate limit because DD_TRACE_RATE_LIMIT is invalid: %v", err)
+		return defaultLimiter
+	}
+	if l < 0.0 {
+		log.Warn("using default rate limit because DD_TRACE_RATE_LIMIT is negative: %f", l)
+		return defaultLimiter
+	}
+	return rate.NewLimiter(rate.Limit(l), int(math.Ceil(r*l)))
```

Comment:
The second argument is confusing. I think it's fine to set it to `1` since we'll only ever be calling `AllowN` with `1`: https://github.com/golang/time/blob/master/rate/rate.go#L74-L77. Then we can also get rid of the argument `r`.